tests/test_client_session.py

```diff
@@ -0,0 +1,235 @@
+                     ("h2", "header2"),
+                     ("h3", "header3")],
+            loop=self.loop)
+        self.assertEqual(
+            set(session._default_headers),
+            set([("h1", "header1"),
+                 ("h2", "header2"),
+                 ("h3", "header3")]))
+        # Test MultiDict
+        session = ClientSession(
+            headers=MultiDict(
+                [("h1", "header1"),
+                 ("h2", "header2"),
+                 ("h3", "header3")]),
+            loop=self.loop)
+        self.assertEqual(
+            set(session._default_headers),
+            set([("h1", "header1"),
+                 ("h2", "header2"),
+                 ("h3", "header3")]))
+
+    def test_init_cookies(self):
+        # Test simple dict
+        session = ClientSession(
+            cookies={
+                "c1": "cookie1",
+                "c2": "cookie2"
+            }, loop=self.loop)
+        self.assertEqual(set(session.cookies), {'c1', 'c2'})
+        self.assertEqual(session.cookies['c1'].value, 'cookie1')
+        self.assertEqual(session.cookies['c2'].value, 'cookie2')
+        # Test list of tuples
+        session = ClientSession(
+            cookies=[("c1", "cookie1"),
+                     ("c2", "cookie2")],
+            loop=self.loop)
+        self.assertEqual(set(session.cookies), {'c1', 'c2'})
+        self.assertEqual(session.cookies['c1'].value, 'cookie1')
+        self.assertEqual(session.cookies['c2'].value, 'cookie2')
+
+    def test_merge_headers(self):
+        # Check incoming simple dict
+        session = ClientSession(
+            headers={
+                "h1": "header1",
+                "h2": "header2"
+            }, loop=self.loop)
+        headers = session._prepare_headers({
+            "h1": "h1"
+        })
+        self.assertIsInstance(headers, CIMultiDict)
+        self.assertEqual(headers, CIMultiDict([
+            ("h1", "h1"),
+            ("h2", "header2")
+        ]))
+        # Check incoming multi dict
+        session = ClientSession(
+            headers={
+                "h1": "header1",
+                "h2": "header2"
+            }, loop=self.loop)
+        headers = session._prepare_headers(MultiDict([("h1", "h1")]))
+        self.assertIsInstance(headers, CIMultiDict)
+        self.assertEqual(headers, CIMultiDict([
+            ("h1", "h1"),
+            ("h2", "header2")
+        ]))
+        # Check incoming list of tuples
+        session = ClientSession(
+            headers={
+                "h1": "header1",
+                "h2": "header2"
+            }, loop=self.loop)
+        headers = session._prepare_headers([("h1", "h1")])
+        self.assertIsInstance(headers, CIMultiDict)
+        self.assertEqual(headers, CIMultiDict([
+            ("h1", "h1"),
+            ("h2", "header2")
+        ]))
+
+    @unittest.mock.patch("aiohttp.client.ClientSession.request")
+    def test_http_methods(self, patched):
```

Comment:
This test method shouldn't be SO long. Please split it into more trivial cases which would help to see what exactly is broken.